File: Algorithms/Sorting/MergeSort.java
public class MergeSort{

	private static int[] numbers={2,10,4,7,3,5,12,8,1,6,3};

	private static int[] helper;

	public static void main(String args[]){


		System.out.println("Array before sorting : ");

		for(int i=0;i<numbers.length;i++){
			System.out.print(numbers[i]+" ");
		}

		//Sorting starts

		sort();

		//Sorting ends


		System.out.println("\n\nArray after sorting : ");

		for(int i=0;i<numbers.length;i++){
			System.out.print(numbers[i]+" ");
		}

		System.out.println();

	}

	public static void sort(){
		int low=0;
		int high=numbers.length-1;
		helper=new int[high+1];
		mergeSort(low, high);
	}

	public static void mergeSort(int low, int high){
		if(low>=high){
			return;
		}
		int middle=low+(high-low)/2;
		mergeSort(low,middle);
		mergeSort(middle+1, high);
		merge(low, middle, high);
	}

	public static void merge(int low, int middle, int high){
		for(int i=low;i<=high;i++){
			helper[i]=numbers[i];
		}

		int index=low;	
		int lower=low;
		int upper=middle+1;
		while(lower<=middle && upper<=high){
			if(helper[lower]<helper[upper]){
				numbers[index++]=helper[lower++];
			}
			else{
				numbers[index++]=helper[upper++];
			}
		}
		while(lower<=middle){
			numbers[index++]=helper[lower++];
		}
	}
}

File: Algorithms/Sorting/merge_sort.py
"""
Is merge sort In Place? I guess yes

"""

def sort(array, start, end):
    if(start == end):
        return
    mid=int((start+end)/2)
    sort(array, start, mid)
    sort(array, mid+1, end)
    merge(array, start, mid + 1, end)

def merge(array, start, mid, end):
    helper=[None] * len(array)
    helper[start:end +1] = array[start:end+1]
    arrayIndex=start
    fixedMid=mid # This indicate index of first element of second array

    while(start < fixedMid and mid <= end):
        if(helper[start] <= helper[mid]):
            array[arrayIndex] = helper[start]
            arrayIndex = arrayIndex + 1
            start = start + 1
        else:
            array[arrayIndex] = helper[mid]
            arrayIndex = arrayIndex + 1
            mid = mid + 1
    while(start < fixedMid):
            array[arrayIndex] = helper[start]
            arrayIndex = arrayIndex + 1
            start = start + 1
    while(mid <= end):
            array[arrayIndex] = helper[mid]
            arrayIndex = arrayIndex + 1
            mid = mid + 1


array = [32,46,77,4344564,7322,3,46,7,32457,7542,4,667,54]
print "Array Befor Sorting"
print array
print "\n\n"
sort(array, 0, len(array)-1) #first index 0, last index len -1
print "Array After Sorting"
print array
print "\n\n"
